libraries/AP_SmartAudio/AP_SmartAudio.cpp

```diff
@@ -0,0 +1,578 @@
+            }
+
+            if (requests_queue.pop(current_command)) {
+                // send the popped command from bugger
+                send_request(current_command.frame, current_command.frame_size);
+
+                now = AP_HAL::millis();
+                // it takes roughly 15ms to send a request, don't turn around and try and read until
+                // this time has elapsed
+                hal.scheduler->delay(20);
+
+                _last_request_sent_ms = now;
+
+                // next loop we expect a response
+                _is_waiting_response = true;
+            }
+        }
+
+        // nothing going on so give CPU to someone else
+        if (!_is_waiting_response && !_initialised) {
+            hal.scheduler->delay(500);
+        }
+
+        // On my Unify Pro32 the SmartAudio response is sent exactly 100ms after the request
+        // and the initial response is 40ms long so we should wait at least 140ms before giving up
+        if (now - _last_request_sent_ms < 200 && _is_waiting_response) {
+
+            // setup sheduler delay to 50 ms again after response processes
+            if (!read_response(_response_buffer)) {
+                hal.scheduler->delay(10);
+                res_retries++;
+            } else {
+                res_retries = 0;
+            }
+
+        } else if (_is_waiting_response) { // timeout
+            // process autobaud routine
+            update_baud_rate();
+            _port->discard_input();
+            _inline_buffer_length = 0;
+            _is_waiting_response = false;
+        } else {
+            if (AP::vtx().have_params_changed() ||_vtx_power_change_pending
+                || _vtx_freq_change_pending || _vtx_options_change_pending) {
+                update_vtx_params();
+                // we've tried to udpate something, re-request the settings so that they
+                // are reflected correctly
+                request_settings();
+            }
+        }
+    }
+}
+
+// send requests to the VTX to match the configured VTX parameters
+void AP_SmartAudio::update_vtx_params()
+{
+    AP_VideoTX& vtx = AP::vtx();
+
+    _vtx_freq_change_pending = vtx.update_band() || vtx.update_channel() || vtx.update_frequency() || _vtx_freq_change_pending;
+    _vtx_power_change_pending = vtx.update_power() || _vtx_power_change_pending;
+    _vtx_options_change_pending = vtx.update_options() || _vtx_options_change_pending;
+
+    if (_vtx_freq_change_pending || _vtx_power_change_pending || _vtx_options_change_pending) {
+        // make the desired frequency match the desired band and channel
+        if (_vtx_freq_change_pending) {
+            if (vtx.update_band() || vtx.update_channel()) {
+                vtx.update_configured_frequency();
+            } else {
+                vtx.update_configured_channel_and_band();
+            }
+        }
+
+        debug("update_params(): freq %d->%d, chan: %d->%d, band: %d->%d, pwr: %d->%d, opts: %d->%d",
```

Comment:
This is triggering every 5 seconds for me, even though the conditions above seem to suggest that it should only happen when there is a change. Is this intentional?

Reply:
Yeah, so this is one of the issues I have to resolve. If you pick a combination of settings that aren't allowed by the VTX it will keep trying to set them even though the VTX has replied with something different to what you set. Common ones are picking a frequency that is only allowed if you unlock.

Reply:
Fixed now I believe